sdk/communication/Shared/src/CommunicationIdentifierSerializer.cs

```diff
@@ -11,12 +11,45 @@ internal class CommunicationIdentifierSerializer
     {
         public static CommunicationIdentifier Deserialize(CommunicationIdentifierModel identifier)
         {
-           string rawId = AssertNotNull(identifier.RawId, nameof(identifier.RawId), nameof(CommunicationIdentifierModel));
+            string rawId = AssertNotNull(identifier.RawId, nameof(identifier.RawId), nameof(CommunicationIdentifierModel));
 
             AssertMaximumOneNestedModel(identifier);
 
-            if (identifier.CommunicationUser is CommunicationUserIdentifierModel user)
-                return new CommunicationUserIdentifier(AssertNotNull(user.Id, nameof(user.Id), nameof(CommunicationUserIdentifierModel)));
+            if (identifier.Kind is not null)
+            {
+                if (identifier.Kind == CommunicationIdentifierModelKind.CommunicationUser
+                    && identifier.CommunicationUser is not null)
+                {
+                    return new CommunicationUserIdentifier(AssertNotNull(identifier.CommunicationUser.Id, nameof(identifier.CommunicationUser.Id), nameof(CommunicationUserIdentifierModel)));
+                }
+
+                if (identifier.Kind == CommunicationIdentifierModelKind.PhoneNumber
+                    && identifier.PhoneNumber is not null)
+                {
+                    return new PhoneNumberIdentifier(
+                        AssertNotNull(identifier.PhoneNumber.Value, nameof(identifier.PhoneNumber.Value), nameof(PhoneNumberIdentifierModel)),
+                        AssertNotNull(identifier.RawId, nameof(identifier.RawId), nameof(PhoneNumberIdentifierModel)));
+                }
+
+                if (identifier.Kind == CommunicationIdentifierModelKind.MicrosoftTeamsUser
+                    && identifier.MicrosoftTeamsUser is not null)
+                {
+                    var user = identifier.MicrosoftTeamsUser;
+                    return new MicrosoftTeamsUserIdentifier(
+                          AssertNotNull(user.UserId, nameof(user.UserId), nameof(MicrosoftTeamsUserIdentifierModel)),
+                          AssertNotNull(user.IsAnonymous, nameof(user.IsAnonymous), nameof(MicrosoftTeamsUserIdentifierModel)),
+                          Deserialize(AssertNotNull(user.Cloud, nameof(user.Cloud), nameof(MicrosoftTeamsUserIdentifierModel))),
+                          rawId);
+                }
+
+                return new UnknownIdentifier(rawId);
```

Comment:
We should probably replace all occurrences of `return new UnknownIdentifier(rawId);` with `return CommunicationIdentifier.FromRawId(rawId)` here and adjust the test accordingly.

Reply:
Reasoning: It's really weird that we're expecting an `UnknownIdentifier` when we've specified the `kind`.
![image](https://user-images.githubusercontent.com/9810625/196921851-d76b2ac1-65f8-498b-b51d-10af689ead43.png)

Reply:
totally agree, good point. Even if the other variables are not set we can identify the identifier type from the rawID, btw can we apply also here my comments from java(https://github.com/Azure/azure-sdk-for-java/pull/31522#discussion_r999458918, https://github.com/Azure/azure-sdk-for-java/pull/31522#discussion_r998399746)

Reply:
Ok, after the internal discussion I gave it some more thought and here's how I'd design the behaviors:

-	RawId missing – fail as it’s a required property in responses (as specified in the swagger). We do fail now -> OK.
-	IdentifierModel missing entirely OR not matching the kind property (if specified) - The polymorphic type is broken, we don’t know what’s going on (probably a service returns something unexpected). Let’s not try to be smart and use UnknownIdentifier
-	RawId not matching IdentifierModel  - We haven’t done these checks before, we probably shouldn’t add them now either. If there is a problem with a service, the error will bubble up either way. We could try to fail early but if we add a new MRI format for, for instance, PhoneNumber in the future, this validation would prevent using an older version of an SDK. It’s not even clear how we should do it – deserialize through both MRI and IdentifierModel and then compare?...
-	Kind is not present, IdentifierModel is missing, RawID is present – this probably means that the IdentifierModel has not been recognized (we might have added a new model in the service that’s not yet understood by an older version of the SDK). It’s unlikely that FromRawId will be able to recognize the model either.

So, for the most part, the PR is fine. But to distil some changes from all this:
- I'd get rid of the ` return new UnknownIdentifier(rawId);` -> it's redundant
- I'd also refactor the logic on lines 48-68 to extract the "kind" and use it as a secondary input for the `kind` variable (if kind is not defined in the response). this will hugely reduce the code duplication
```c#
var kind = identifier.Kind ?? GetKind(identifier);
if (kind is not null)
```

Reply:
@petrsvihlik please see PR https://github.com/Azure/azure-sdk-for-net/pull/31913